### ISSUE TITLE
detection not working on BCHW format

### ISSUE THREAD

**Issue by kgonia** [ultralytics/ultralytics#3352]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection

### Bug

same image in RGB format loaded by PIL returns 7 boxes. After convertion to torch BCHW format I got zero boxes.

### Environment

YOLO: Ultralytics YOLOv8.0.121  Python-3.9.7 torch-2.0.1+cu118 CUDA:0 (NVIDIA GeForce RTX 2070 SUPER, 8192MiB)
OS: Windows 10
Python: 3.9

### Minimal Reproducible Example

import urllib

import torch
from PIL import Image
from torchvision.transforms import Compose, Resize, ToTensor, PILToTensor
from ultralytics import YOLO

transform = Compose([
    Resize((640, 640)),
    ToTensor()
])

yolo_model = YOLO("yolo/yolov8n-face.pt")

urllib.request.urlretrieve(
    'https://cf.ltkcdn.net/family/images/orig/200821-2121x1414-family.jpg',
    "family.jpg")

img = Image.open("family.jpg").convert('RGB')

img_t = transform(img)
img_t = torch.stack([img_t])

result = yolo_model(img)
result_t = yolo_model(img_t)

print(len(result[0].boxes) == 7)
print(len(result_t[0].boxes) == 0)

### Additional

I'm using **yolov8n-face.pt** model from https://github.com/akanametov/yolov8-face

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by kgonia** [comment]
same situation with official model

```
import urllib

import torch
from PIL import Image
from torchvision.transforms import Compose, Resize, ToTensor, PILToTensor
from ultralytics import YOLO

transform = Compose([
    Resize((1024, 1024)),
    ToTensor()
])

yolo_model = YOLO("yolov8n.pt")

urllib.request.urlretrieve(
    'https://ultralytics.com/images/bus.jpg',
    "bus.jpg")

img = Image.open("bus.jpg").convert('RGB')

img_t = transform(img)
img_t = torch.stack([img_t])

result = yolo_model(img)
result_t = yolo_model(img_t)

print(len(result[0].boxes) == 6)
print(len(result_t[0].boxes) == 0)
```

**Comment by kgonia** [comment]
@glenn-jocher 

> The detection mode expects input in the HWC format

BCHW format is also compatible according to the official documentation, moreover BCHW is used internaly. All other formats are converted. I already propose changes in code to handle this case when BCHW is arleady provided. 
The prevalent practice when using PyTorch involves using the ToTensor() operation. It's reasonable to expect this operation has already been performed.

> You can do this by omitting the conversion to ToTensor() in your transform pipeline

Thank you for your suggestion. However, the use of ToTensor() operation in this context is not arbitrary, it's a crucial part of our data transformation pipeline in the full training script.

The provided code snippet is a simplified example to demonstrate the issue, but in the complete training scenario, the ToTensor() operation is an integral part of the Dataset class transform operation.

Using ToTensor() in data transformations is a common practice in PyTorch. It not only converts PIL Images or NumPy ndarrays into tensors, but also scales the pixel intensities to the range of [0, 1], a commonly used normalization step in image processing tasks. This helps improve model performance and the stability of the training process.

Therefore, while it's true that we could omit ToTensor() in this minimal example, in our full script, removing it would disrupt the standard flow of data preparation. By maintaining the ToTensor() operation, we ensure that our pipeline is consistent with the common practices within the PyTorch ecosystem.

**Comment by glenn-jocher** [comment]
@kgonia 

Thank you for your input. I understand that the ToTensor() operation is an important part of your data transformation pipeline and is necessary for your complete training scenario. Please accept my apologies for not considering the full context of your code snippet.

You are correct that the BCHW format is compatible with the YOLOv8 detection mode, and it is used internally in the model. The model can handle different input formats, including BCHW, and any other formats are converted accordingly.

In the specific code snippet you provided, the conversion to BCHW format after applying the ToTensor() operation seems to be causing the issue. It is expected that the detection should still work with the BCHW format in this case.

While my previous suggestion to omit the ToTensor() operation in the transform pipeline would solve the issue in this minimal example, I now understand that it would disrupt the standard flow of your data preparation in the full script.

To summarize, it seems that there might be an issue with the YOLOv8 detection mode when working with the BCHW format after the ToTensor() operation is applied. I'll need to investigate this further to provide a more accurate solution.

Thank you for bringing this to our attention, and I apologize for any confusion caused. I will continue researching this issue and provide an update as soon as possible.

Please let me know if you have any further questions or if there's anything else I can assist you with.

Best regards,
Glenn Jocher

**Comment by kgonia** [comment]
@glenn-jocher please consider my PR https://github.com/ultralytics/ultralytics/pull/3354 as solution. I see normalization was introduced as part of SAM implementation https://github.com/ultralytics/ultralytics/pull/2298 so I'm not sure if my proposition won't brake any other workflow but there is no coverage in tests for other cases and ToTensor() is prevalent practise.

**Comment by glenn-jocher** [comment]
@kgonia thank you for your PR and for bringing up the issue. We appreciate your contribution to the YOLOv8 repository.

The normalization step implemented in the original code was introduced as part of the SAM implementation, and it seems that removing it might affect other workflows. We don't have coverage in tests for other cases, and the ToTensor() operation is a common practice.

We will review your PR and consider its impact on other workflows to ensure it doesn't introduce any unintended issues. We appreciate your understanding of our process and your patience as we evaluate the proposed solution.

Thank you once again for your contribution, and we will keep you updated on any progress regarding your PR.

Best regards,
Glenn Jocher

**Comment by glenn-jocher** [comment]
@kgonia I've merged your PR into a larger 8.0.122 release that will include a fix for Tensor inference to allow ToTensor() ops. I'm using this script to test the branch:

```
from PIL import Image
from ultralytics import YOLO
from ultralytics.yolo.utils.downloads import download
import torch
import numpy as np
from torchvision.transforms import ToTensor

# Load a model
model = YOLO('yolov8n.pt')  # pretrained YOLOv8n model

# Open an image using PIL
download('https://ultralytics/images/bus.jpg')
source = Image.open('bus.jpg').resize((800,1088))  # width, height

# Run inference on the source
results = model(source)  # list of Results objects
results = model(np.array(source))  # list of Results objects
results = model(torch.from_numpy(np.array(source)).permute(2,0,1).unsqueeze(0) / 255.0)  # list of Results objects
results = model(ToTensor()(np.array(source)))  # list of Results objects
```

**Comment by Sir-Photch** [comment]
The current implementation is not correct. If a tensor with batch size > 1 is passed as an argument, the result array includes the same images when calling `results[i].plot()`.

**Comment by glenn-jocher** [comment]
@Sir-Photch i appreciate your feedback and for bringing this to our attention. I understand the problem you're facing. If I get it right, you're saying that when you pass a batch of images (in your example, `imageA` and `imageB`) through the model and plot the results for each image, they're identical, which isn't supposed to be the case.

This could be a potential issue with how the model processes batches of images or how the results are stored after doing the detection work. It seems as if the model is applying the same emission of bounding boxes for all images in the batch instead of applying unique detections for each image.

We will further investigate this issue. Thank you very much for providing this insightful description, it is very helpful for us to improve the YOLOv8 model.
